nixos/syncoid: change sshKey options to disallow path-types

tejing1 opened the pull request:
The previous behavior obfuscates the default behaviors of the nix language with regard to strings and paths. Better to give an error and let them learn the correct way to do what they wanted than to silently fix their mistake and enhance the misunderstanding.

## Description of changes

Limit `sshKey` options to string types rather than using `toString` to prevent path-types from being copied to the store.

Users who try to pass a path type are prevented from making the fairly horrific mistake of copying their ssh keys to the nix store, just as before, but now the mistake is not reinforced as correct code in their minds. Let's not confuse the path/string issue any more than it already is.

Anyone who wants to do something unusual can always use a string antiquote in their own code when setting the option, so no functionality is lost.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tejing1 committed:
nixos/syncoid: change sshKey options to disallow path-types

The previous behavior obfuscates the default behaviors of the nix
language with regard to strings and paths. Better to give an error and
let them learn the correct way to do what they wanted than to silently
fix their mistake and enhance the misunderstanding.

tejing1 commented:
I suppose this is worthy of an item in the release notes. I'll add it.

JulienMalka reviewed (commented):
I we are making a breaking change I'd even change the option names to `privateSshKeyFile` or something like that. A lot of modules have `*file` as name for options that represent a path to a secret. Also, changing the option to accept a string, while maybe more correct because it doesn't involve a transformation, is probably less clear for the user trying to understand what the option does. At least, the type path conveyed the meaning that the user should enter the path to a file here.

tejing1 commented:
> maybe more correct because it doesn't involve a transformation

Not sure if we're quite on the same page about why I want this. It's a pedagogical issue for users, rather than anything about transformation as such. I don't want users supplying path-types to other options and being extra surprised and confused later when it copies to the nix store. After all, they already "found out it doesn't do that" from the behavior of this option. Consistency in behavior, and ease of learning the ecosystem's patterns, is what I'm going for.

> probably less clear for the user trying to understand what the option does. At least, the type path conveyed the meaning that the user should enter the path to a file here.

This can be fixed by altering the type itself to have a different `description`, if we feel it's important.

JulienMalka commented:
> Not sure if we're quite on the same page about why I want this. It's a pedagogical issue for users, rather than anything about transformation as such. I don't want users supplying path-types to other options and being extra surprised and confused later when it copies to the nix store. After all, they already "found out it doesn't do that" from the behavior of this option. Consistency in behavior, and ease of learning the ecosystem's patterns, is what I'm going for.

I do understand why you want this change and why this brings more consistency to the behavior of the `types.path` type. However, I also think this makes us loose quite a lot in terms of readability. The type of the option is displayed on `search.nixos.org` and is a piece of information the user will read in order to understand what the option does. With an `str` type, the user may think they have to paste their ssh key here which is arguably worse. Additionally, types are used to validate the data entered by the user. While Nix would give a type error in case the user did not provide a path, with this change the error would happen at runtime and probably be harder to debug.

tejing1 commented:
Alright, fair enough. I'll change it to use a custom type that does the same checks `types.path` does on the string, and self-describes as `string containing path`, or something along those lines, for nixos option search, so the type would show as `null or string containing path`.

tejing1 commented:
Updated to fix merge conflict with #252747, and in light of the changes made there, I'm reverting to just `nullOr str`. The option description does say it's supposed to be a file, and I added "Path to the" there, just for clarity.
